File: app/src/main/java/com/example/prashant/databinding/ui/ListItemFragment.java
package com.example.prashant.databinding.ui;

import android.arch.lifecycle.ViewModelProvider;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.databinding.DataBindingUtil;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.prashant.databinding.R;
import com.example.prashant.databinding.data.Contact;
import com.example.prashant.databinding.databinding.FragmentListItemBinding;
import com.example.prashant.databinding.observer.ItemFactory;
import com.example.prashant.databinding.observer.ItemViewModel;

import java.util.ArrayList;
import java.util.Objects;

import javax.inject.Inject;

import dagger.android.support.AndroidSupportInjection;

/**
 * A simple {@link Fragment} subclass.
 * Activities that contain this fragment must implement the
 * {@link ListItemFragment.OnFragmentInteractionListener} interface
 * to handle interaction events.
 * Use the {@link ListItemFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class ListItemFragment extends Fragment {
    // TODO: Rename parameter arguments, choose names that match
    // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
    private static final String ARG_PARAM1 = "param1";

    // TODO: Rename and change types of parameters
    private String mParam1;

    private OnFragmentInteractionListener mListener;

    private ArrayList<Contact> contactArrayList = new ArrayList<>();

    @Inject
    public ViewModelProvider.Factory itemFactory;

    private ItemViewModel model;

    public ListItemFragment() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param param1 Parameter 1.
     * @return A new instance of fragment ListItemFragment.
     */
    // TODO: Rename and change types and number of parameters
    public static ListItemFragment newInstance(String param1) {
        ListItemFragment fragment = new ListItemFragment();
        Bundle args = new Bundle();
        args.putString(ARG_PARAM1, param1);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mParam1 = getArguments().getString(ARG_PARAM1);
        }

        // Create a ViewModel the first time the system calls an activity's onCreate() method.
        // Re-created activities receive the same MyViewModel instance created by the first activity.
        model = ViewModelProviders.of(this, itemFactory).get(ItemViewModel.class);
        model.init();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        FragmentListItemBinding binding = DataBindingUtil.inflate(
                inflater, R.layout.fragment_list_item, container, false);

        RecyclerView recyclerView = binding.recyclerView;
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        ItemAdapter adapter = new ItemAdapter(contactArrayList);
        recyclerView.setAdapter(adapter);

        model.getContacts().observe(this, list -> {
            // Update UI
            contactArrayList.clear();
            contactArrayList.addAll(list);
        });

        return binding.getRoot();
    }

    // TODO: Rename method, update argument and hook method into UI event
    public void onButtonPressed(Uri uri) {
        if (mListener != null) {
            mListener.onFragmentInteraction(uri);
        }
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        // config dagger
        AndroidSupportInjection.inject(this);

//        if (context instanceof OnFragmentInteractionListener) {
//            mListener = (OnFragmentInteractionListener) context;
//        } else {
//            throw new RuntimeException(context.toString()
//                    + " must implement OnFragmentInteractionListener");
//        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     * <p>
     * See the Android Training lesson <a href=
     * "http://developer.android.com/training/basics/fragments/communicating.html"
     * >Communicating with Other Fragments</a> for more information.
     */
    public interface OnFragmentInteractionListener {
        // TODO: Update argument type and name
        void onFragmentInteraction(Uri uri);
    }
}

File: app/src/main/java/com/example/prashant/databinding/ui/ListItemFragment.kt
package com.example.prashant.databinding.ui

import android.Manifest
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingComponent
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import com.example.prashant.databinding.R
import com.example.prashant.databinding.binding.FragmentDataBindingComponent
import com.example.prashant.databinding.databinding.ListBinding
import com.example.prashant.databinding.di.Injectable
import com.example.prashant.databinding.observer.ItemViewModel
import com.example.prashant.databinding.utils.extensionUtil.AppExecutors
import com.example.prashant.databinding.utils.extensionUtil.autoCleared
import com.example.prashant.databinding.utils.permissionUtil.BaseFragment
import com.google.android.material.snackbar.Snackbar
import javax.inject.Inject

/**
 * A simple [Fragment] subclass that extends base fragment.
 * Activities that contain this fragment must implement the
 * [ListItemFragment.OnFragmentInteractionListener] interface
 * to handle interaction events.
 * Use the [ListItemFragment.newInstance] factory method to
 * create an instance of this fragment.
 */
class ListItemFragment : BaseFragment(), Injectable {

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory

    @Inject
    lateinit var viewModel: ItemViewModel

    @Inject
    lateinit var executors: AppExecutors

    var binding by autoCleared<ListBinding>()

    var dataBindingComponent: DataBindingComponent = FragmentDataBindingComponent(this)

    lateinit var mAdapter: ItemAdapter

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        binding = DataBindingUtil.inflate(
                inflater,
                R.layout.fragment_list_item,
                container,
                false,
                dataBindingComponent)
        return binding.root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        viewModel = ViewModelProviders
                .of(this, viewModelFactory)
                .get(ItemViewModel::class.java)

        mAdapter = ItemAdapter(dataBindingComponent, executors, null)

        binding.fab.setOnClickListener {
            requestCall()
        }

        binding.let {
            it.rvList.adapter = mAdapter
            it.lifecycleOwner = this
        }

        viewModel.let {
            it.init()
            it.contacts.observe(this, Observer { list ->
                mAdapter.submitList(list)
            })
        }
    }

    private fun requestCall() {
        val permissionList = listOf(Manifest.permission.CALL_PHONE,
                Manifest.permission.WRITE_EXTERNAL_STORAGE)
        requestThenRun(
                permissions = permissionList,
                action = this::showSuccess,
                failAction = this::showError)
    }

    private fun showSuccess() {
        showSnackBar("Your permission request accepted")
    }

    private fun showError(list: List<String>) {
        showSnackBar(list.toString())
    }

    private fun showSnackBar(message: String) {
        Snackbar.make(binding.root, message, Snackbar.LENGTH_SHORT).show()
    }
}
